### ISSUE TITLE
Bounty for Native 15khz CRT Support in Retroarch [$50]

### ISSUE THREAD

**Issue by asimonf** [libretro/RetroArch#4879]
### Description

This bounty is for adding native 15khz support into RetroArch so that it supports displaying a game's native resolution and refresh rate in a 15khz CRT Monitor or TV similar to how GroovyMame actually supports resolution switching and setting on CRT monitors depending on the game or machine.

### Background

Older consoles usually supported more than one resolution for their games, horizontal resolution isn't usually a problem and this is currently solved in Retroarch using superwide modes (2560x240 or 3840x240, for example). Continuing to use superwide modes is acceptable for horizontal resolution. However, some games/consoles actually support multiple vertical resolution or modes.

One example of the above is the PSX that uses 480i in the BIOS and then makes the transition to whatever mode the game uses, which could be 640x240p, 640x480i (assuming NTSC games). This second situation is currently resolved for 31Khz monitors by using a shader that imitates interlacing on games that send more than 400 vertical lines. But in 15khz monitors or TVs, this is not an option, so native mode switching is required.

Another situation is that PAL games actually ran at a vertical refresh rate of 50hz. Speeding the game to 60hz is one option, but switching to a 50hz mode natively on a CRT monitor would be preferable. 

Feedback is welcome.

**Comment by bparker06** [comment]
RetroArch already supports native resolution switching when in fullscreen mode using the config options video_fullscreen_x and y. The main issue is that 90% of modern graphics cards cannot support a slow enough pixel clock for a native 240p resolution. The way video decoder ASICs used in modern TVs handle it is with the wide horizontal frame size like you mentioned, and this is part of the CEA-861 standard. One exception to this is a very recent rpi firmware that enables native 240p over composite.

Basically what I'm saying is that I'm not aware of anything RetroArch could do to further solve this problem.

**Comment by asimonf** [comment]
Some games and consoles actually switch modes inbetween (mostly between interlaced and progressive modes). Horizontal resolution switching can be solved with wide resolutions (The bsnes accurate mode core actually renders internally at 512px horizontally to handle switching between hires and lores mode on the SNES, if I'm not mistaken), but there is currently no vertical mode switching.

**Comment by hizzlekizzle** [comment]
@bparker06 
I think the idea would be to have the modes already exist and then set a mask for the video_fullscreen_y value (presumably something like 0 or -1 or whatever) that would mean "get the vertical res from the core and try to use that". It's up to the user how to get a 15 khz signal out of their GPU (usually by using hacked-up old ATi drivers, IIRC). This is how GroovyMAME does it, I think.

@asimonf 
You can use 2x integer scale with the misc/interlacing shader and it should show solid black lines on half of the fields for non-interlaced content and then alternate fields on interlaced content. This is for 31 khz displays, though.

**Comment by asimonf** [comment]
@hizzlekizzle 
Using a mask for the vertical resolution would probably be a good idea. I understand that Switchres does something of the like when using wide resolutions by calculating the closest mode available to the requested mode by the game when using GroovyMame. The only other part of the equation would probably be refresh rate to make sure PAL games use 50hz and NTSC games use 60hz. 

The part of actually generating the 15khz signal to drive the monitor/tv is left as an exercise to the user. BTW, new hacked drivers have been released that support more modern AMD cards. Actually, the new hacked drivers support cards all the way up to Fury gen, so the only unsupported cards are the Polaris-based ones like the RX 480, RX580 and company.

I already have a setup ready, so I'd be more than happy to help debug anything that arises.

**Comment by aliaspider** [comment]
seamless switching between 480i and 240p isn't really possible on PCs, even when using the same pixel clock, you would always end up with a monitor re-sync, unlike a real psx which can just toggle double strike mode on the fly (the high res menu in chrono chross for example).

**Comment by Kivutar** [comment]
Maybe we could do this for Raspberry Pi, if not possible for PCs. I see that they are used a lot in arcade cabinets now that the firmware is able to do 240p.

This file can give you an idea on how the timing is configured for each platform https://github.com/mortaca/RGB-Pi/blob/master/HDMI_Timings.txt

For arcade games, it is way more complex than what we see in this file. And I doubt that RetroArch can cleanly detect if we're using a 50 or 60hz CRT.

But we could do something like this: having a hdmi_timing entry in the retroarch.cfg, and people would use it in core/game overrides when a game is launched, in addition to a custom viewport. This would help for the arcade cabinet use case at least.

**Comment by KingK76** [comment]
I don't see the issue with res switching and a 15Khz Arcade CRT.  In MAME my hardware setup works perfectly fine.  I recently purchased a 2TB Hyperspin setup and when MAME every games works great on my 15KHz CRT with a JPAC and an old AMD 5450 GPU.  No matter what game or resolution everything just works.  But it is Retroarch that I am currently having issues with.  Every system emulated through Retroarch gives my a messed up double image on the screen.  Is this due to the issue the OP wants fixed?  I will attach a photo of my CRT when Retroarch is running the emulation.  It sucks and I would love for this to be fixed!!!

![double screen retroarch issue 2](https://user-images.githubusercontent.com/30483517/28649767-3dba2812-7245-11e7-95fa-958c2eec3d04.jpg)

**Comment by twinaphex** [comment]
We don't own the hardware you're using so it's impossible to tell really and impossible to cater to. At best we could plug the bounty a bit more and see if there's any interest by a developer who would be able to do this.

**Comment by twinaphex** [comment]
Hmm, the bounty is closed and there is no $50 attached to it anymore. Can't raise awareness to this bounty then.

**Comment by hizzlekizzle** [comment]
@KingK76 No problems here on my cabinet. If you have any "super" resolutions already set up, like 1920 or 3840 by 240/224, you can try setting those values in your retroarch.cfg's video_fullscreen_x and video_fullscreen_y lines.

**Comment by KingK76** [comment]
@hizzlekizzle Interesting...  I will try that!  Man if that works I should just send you the $50!  lol...

**Comment by KingK76** [comment]
@hizzlekizzle Hey man thanks for the suggestion but it didn't work...  You did however inspire to start farting around enough to see if I couldn't fix the issue.  And guess what?  Success!  Well sort of...  I now have a single centered image on the screen.  The problem is that it is only filling around 80% of the screen.  Anyway the solution was to change the Video driver from "gl" to "d3d".  I had to do the exact same thing on the Nintendo64 emulator I am running in the same unit.  There are still some emus that are doing the double image thing but I can't find a way to change the video plugin or "driver" to Direct3D.  Like I also have a Turbografix 16 emulator that is still doing the double image crap...  I wanted to thank you for your suggestion though, it really did get me to try harder to figure the issue out.  BTW...  Do you have any ideas for getting the image to fill the screen in Retroarch now that I have a single centered image???

Thanks Again

Kris

**Comment by hizzlekizzle** [comment]
What kind of resolution are you running? If RA is taking up less than the full screen, it's probably a result of integer scaling being enabled, which is typically a good thing to avoid weird scaling issues. You can disable integer scaling and it *should* fill up the screen, but it's probably not going to look very good. A better option is to get your GPU to go ahead and output the resolution you need, which is going to be either 240 with visible overscan or 224 with no overscan (either way, you may need to go into settings > video and turn overscan cropping on/off and/or mess with core options to get the image to fit properly).

**Comment by Kivutar** [comment]
I was able to get 15Khz working fine on my CRT. Using a raspberry pi 3 and the RGB-Pi scart adapter.

I had to write a custom modeline on a per game basis however. And set the resolution to 1920x224 or something close.

And I didn't went as far as being able to disable frame throtelling and video buffering, which is the holy grail of this kind of work, if I understand well.

I don't think there is anything to add to RA to do better. It's more a matter of finding the right modelines per game, and using per game configs, and studying the matter un detail.
